qCC/ccCommandLineCommands.cpp

```diff
@@ -2688,6 +2690,164 @@ bool CommandMergeClouds::process(ccCommandLineInterface& cmd)
 	return true;
 }
 
+CommandSetGlobalShift::CommandSetGlobalShift()
+	: ccCommandLineInterface::Command(QObject::tr("Set global shift"), COMMAND_SET_GLOBAL_SHIFT)
+{}
+
+bool CommandSetGlobalShift::process(ccCommandLineInterface& cmd)
+{
+	cmd.print(QObject::tr("[SET GLOBAL SHIFT]"));
+
+	if (cmd.clouds().empty() && cmd.meshes().empty())
+	{
+		return cmd.error(QObject::tr("No loaded entity! (be sure to open one with \"-%1 [filename]\" before \"-%2\")").arg(COMMAND_OPEN, COMMAND_SET_GLOBAL_SHIFT));
+	}
+
+	//process globalshift options first
+	ccCommandLineInterface::GlobalShiftOptions globalShiftOptions;
+	cmd.processGlobalShiftCommand(globalShiftOptions);
+	//if it is not a valid global shift then an error msg already issued.
+	if (globalShiftOptions.mode != ccCommandLineInterface::GlobalShiftOptions::Mode::CUSTOM_GLOBAL_SHIFT)
+	{
+		return cmd.error(QObject::tr("Global shift must be in the form of three coordinates 'x' 'y' 'z'"));
+	}
+	CCVector3d newShift = globalShiftOptions.customGlobalShift;
+
+	//look for additional parameters
+	bool keepOrigFixed = false;
+	while (!cmd.arguments().empty())
+	{
+		QString argument = cmd.arguments().front();
+		if (ccCommandLineInterface::IsCommand(argument, COMMAND_SET_GLOBAL_SHIFT_KEEP_ORIG_FIXED))
+		{
+			//local option confirmed, pop that from front
+			cmd.arguments().pop_front();
+
+			keepOrigFixed = true;
+			cmd.print(QObject::tr("[%1]").arg(COMMAND_SET_GLOBAL_SHIFT_KEEP_ORIG_FIXED));
+		}
+		else {
+			break;
+		}
+	}
+
+	//create an entity vector
+	size_t nrOfClouds = cmd.clouds().size();
+	size_t nrOfMeshes = cmd.meshes().size();
+	size_t nrOfEntities = nrOfClouds + nrOfMeshes;
+	QVector<QPair<ccShiftedObject*, CLEntityDesc*>> entities;
+
+	//add clouds to the vector
+	for (size_t i = 0; i < nrOfClouds; i++)
+	{
+		QPair<ccShiftedObject*, CLEntityDesc*> entity;
+		entity.first = cmd.clouds()[i].pc;
+		entity.second = &cmd.clouds()[i];
+		entities.append(entity);
+	}
+
+	//add meshes to the vector
+	for (size_t i = 0; i < nrOfMeshes; i++)
+	{
+		QPair<ccShiftedObject*, CLEntityDesc*> entity;
+		bool isLocked = false;
+		ccShiftedObject* shifted = ccHObjectCaster::ToShifted(cmd.meshes()[i].mesh, &isLocked);
+		if (shifted && !isLocked)
+		{
+			entity.first = shifted;
+			entity.second = &cmd.meshes()[i];
+			entities.append(entity);
+		}
+	}
+
+	//process both clouds and meshes
+	for (size_t i = 0; i < nrOfEntities; i++)
+	{
+		QPair<ccShiftedObject*, CLEntityDesc*> entity=entities[i];
+		CLEntityDesc& desc = *entity.second;
+		ccShiftedObject* shiftedObject = entity.first;
+		CCVector3d originalShift = shiftedObject->getGlobalShift();
+		cmd.print(QObject::tr("\t[%4 - %5] Original global shift {%1,%2,%3}")
+			.arg(originalShift.x)
+			.arg(originalShift.y)
+			.arg(originalShift.z)
+			.arg(desc.basename)
+			.arg(shiftedObject->getName()));
+
+		//translate entity to keep the same origin
```

Comment:
the same origin as what exactly? I'm still a little bit confused by this feature...

Reply:
-KEEP_ORIG_FIXED means that you change the global shift, but the result cloud stays the same place (global origin is preserved).

It is more likely could be used to get every element to same global shift, if it was imported with "AUTO" function, I wouldn't use this in real word. I think but it is in the gui version as well.

Reply:
Ah sorry, I was confused by the order of things. Here you are changing the Global Shift after having loaded the clouds...